web/src/db/finders.rs

```diff
@@ -1,35 +1,238 @@
+              .fetch_all(db)
+              .await?;
+        Ok(stations)
+    }
+
+    pub async fn by_place_id(db: &DbPool, place_id: &String) -> Result<StationRecord, DbError> {
+        sqlx::query_as!(
+            StationRecord,
+            "SELECT id, name, uic, latitude, longitude, info_de, info_en, info_es, info_fr, info_it, info_nb, info_nl, info_cs, info_da, info_hu, info_ja, info_ko, info_pl, info_pt, info_ru, info_sv, info_tr, info_zh FROM stations WHERE uic = $1",
+            place_id
+        )
+        .fetch_optional(db)
+        .await?
+        .ok_or(DbError::RecordNotFound(format!(
+            "Could not find station with uic #{}",
+            &place_id
+        )))
+    }
+
+    /// Search for stations near a specific geographic point
+    pub async fn by_position(
+        pool: &DbPool,
+        latitude: f64,
+        longitude: f64,
+    ) -> Result<Vec<StationRecord>, DbError> {
+        println!("Searching stations by position");
+        // First, get a larger set of candidates using a bounding box
+        // This is more efficient for the initial filtering
+        let approx_distance_deg = 1.0; // Roughly 100km at equator
+
+        let query = sqlx::query_as!(
+            StationRecord,
+            r#"
+            SELECT *
+            FROM stations
+            WHERE
+                latitude IS NOT NULL
+                AND longitude IS NOT NULL
+                AND latitude BETWEEN ? - ? AND ? + ?
+                AND longitude BETWEEN ? - ? AND ? + ?
+            "#,
+            latitude,
+            approx_distance_deg,
+            latitude,
+            approx_distance_deg,
+            longitude,
+            approx_distance_deg,
+            longitude,
+            approx_distance_deg
+        );
+
+        // Fetch candidates
+        let db_stations = query.fetch_all(pool).await?;
+        // Convert to GeoPositionFinder, which includes custom fields for calculating distance
+        let mut geo_search_results: Vec<GeoPositionFinder> =
+            db_stations.into_iter().map(Into::into).collect();
+
+        // Calculate result distances and sort
+        for result in &mut geo_search_results {
+            if let (Some(lat), Some(lon)) = (result.station.latitude, result.station.longitude) {
+                // Add a custom field for distance (we'll use this for sorting)
+                result.distance = Some(GeoPositionFinder::haversine_distance(
+                    latitude, longitude, lat, lon,
+                ));
+            }
+        }
+
+        // Sort by distance
+        geo_search_results.sort_by(|a, b| {
+            a.distance
+                .unwrap_or(f64::MAX)
+                .partial_cmp(&b.distance.unwrap_or(f64::MAX))
+                .unwrap()
+        });
+
+        let db_stations: Vec<StationRecord> =
+            geo_search_results.into_iter().map(Into::into).collect();
```

Comment:
I think we can do away with `GeoPositionFinder` (at least the struct part) and simplify this code by moving the calculation of `distance` (and later `relevance_score`) into the sorting block:

```suggestion
    let mut db_stations = query.fetch_all(pool).await?;

    db_stations.sort_by_cached_key(|place| {
        let place_lat = place.latitude.expect("Latitude is not present");
        let place_lon = place.longitude.expect("Longitude is not present");
        let distance = haversine_distance(latitude, longitude, place_lat, place_lon);
        (distance * 10000f64) as i64
    });
```